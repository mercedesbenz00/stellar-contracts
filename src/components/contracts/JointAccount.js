import React from 'react'
import Form from 'react-jsonschema-form'
import {Col, Grid, Panel, Row} from 'react-bootstrap'
import sdk from 'stellar-sdk'

import Contracts from '../../api'
import AccountField from '../AccountField'
import CreateButton from '../CreateButton'
import Receipt from '../Receipt'
import {isSignedIn, withServer, withSigner} from '../../utils'

const HelpPanel = () =>
  <Panel bsStyle="info" header="Help">
    <div>Creates a simple joint account on the Stellar Network.</div>
    <div style={{marginTop: '1em'}}>
      This setup allows any member account to make payments (and other medium
      threshold operations).
    </div>
    <div style={{marginTop: '1em'}}>
      However high threshold operations like changing the list of signers
      requires all parties to sign.
    </div>
    <div style={{marginTop: '1em'}}>
      References:
      <div style={{marginLeft: 10}}>
        <div>
          <a href="https://www.stellar.org/developers/guides/concepts/multi-sig.html#example-2-joint-accounts">
            Joint Accounts
          </a>
        </div>
        <div>
          <a href="https://www.stellar.org/developers/guides/concepts/multi-sig.html">
            Multisig
          </a>
        </div>
      </div>
    </div>
  </Panel>

const schema = {
  title: 'Joint Account',
  description:
    "1) Enter Secret Key of an existing account  OR  2) 'Generate' new account  OR  3) 'Use Signer'",
  type: 'object',
  properties: {
    account: {
      title: 'Account',
      type: 'object',
    },
    members: {
      title: 'Members',
      type: 'array',
      items: {
        type: 'string',
        default: '',
      },
    },
    signer: {type: 'string'},
  },
}

const uiSchema = {
  account: {
    'ui:field': 'account',
  },
  members: {
    'ui:options': {
      orderable: false,
    },
    'ui:help':
      "Enter stellar public keys for all accounts to add as signers. Click the '+' button to add more.",
  },
  signer: {
    'ui:widget': 'hidden',
  },
}

const fields = {
  account: AccountField,
}

class JointAccount extends React.Component {
  formData = {
    account: '',
    members: [''],
  }

  state = {isLoading: false}

  constructor(props) {
    super(props)
    this.formData.signer = props.signer ? props.signer : ''
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.signer !== nextProps.signer)
      this.formData.signer = nextProps.signer
  }

  formValidate = (formData, errors) => {
    console.log(`Validate: ${JSON.stringify(formData)}`)

    // check user is signed in as we need a tx signer
    if (!isSignedIn(formData))
      errors.signer.addError(
        "You must be signed in to create this contract. Click 'Sign In' in the page header."
      )

    if (formData.members.length < 1) {
      errors.members.addError(
        "Provide at least 1 member account. Click '+' to add a member field."
      )
    }

    formData.members.forEach((member, idx) => {
      if (!sdk.StrKey.isValidEd25519PublicKey(member)) {
        errors.members[idx].addError(
          'Account is not a valid ed25519 public key'
        )
      }
      // TODO: check account exists on the network too .. (server.loadAccount)
    })

    return errors
  }

  handleOnSubmit = ({formData}) => {
    this.setState({error: null, isLoading: true})
    const contracts = new Contracts(this.props.server)
    const contract = contracts.jointAccount()
    contract
      .create({
        accountSecret: formData.account.secret,
        members: formData.members,
        signer: this.props.signer,
      })
      .then(receipt => {
        console.log(JSON.stringify(receipt))
        this.setState({isLoading: false, receipt: receipt})
      })
      .catch(err => {
        this.setState({
          isLoading: false,
          error: err.detail ? err.detail : err.message,
        })
      })
  }

  render() {
    return (
      <Grid>
        <Row>
          <Col md={8}>
            <Form
              fields={fields}
              formData={this.formData}
              formContext={{signer: this.props.signer}}
              onSubmit={this.handleOnSubmit}
              schema={schema}
              uiSchema={uiSchema}
              validate={this.formValidate}
            >
              <CreateButton
                errorMsg={
                  this.state.error && typeof this.state.error === 'string'
                    ? this.state.error
                    : ''
                }
                isLoading={this.state.isLoading}
              />
            </Form>
            {this.state.receipt && <Receipt receipt={this.state.receipt} />}
          </Col>
          <Col md={4} style={{marginTop: 20}}>
            <HelpPanel />
          </Col>
        </Row>
      </Grid>
    )
  }
}

export default withServer(withSigner(JointAccount))
