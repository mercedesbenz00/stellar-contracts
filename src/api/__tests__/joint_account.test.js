import sdk from 'stellar-sdk'
import JointAccount from '../joint_account'

const SIGNER = sdk.Keypair.fromSecret(
  'SAP4WQ7W3JS72NGGTJ3X7FM3VGMVI5AIWS5VUOXK4MJECLTOPDUE7VMK'
)
const JOINT_NEW = sdk.Keypair.fromSecret(
  'SABRXC3RQJVVUS22N7H6WGZPTWJ7GBKIZ3PBGX4NWKZZSLC4DWIUP4K3'
)
const MEMBER1 = sdk.Keypair.fromPublicKey(
  'GAJ3VIHUHOHPI7HIJ3DS5HJ5K7CMJ5QVVCO4IXBIG32HGM3WAS3OEOMA'
)
const MEMBER2 = sdk.Keypair.fromPublicKey(
  'GBN4BQAIFUEKYOEYAI4ZEWUK7UW4QVCFVT4VD643L5O7IBCZEES3R6D3'
)
const MEMBER3 = sdk.Keypair.fromPublicKey(
  'GAGXSBXBSJUYCYM623OQQK5P3PXT2BOCKRXU7TVAEOZB5CET2ABJX2T6'
)
const MEMBERS = [MEMBER1.publicKey(), MEMBER2.publicKey(), MEMBER3.publicKey()]

const signerObj = (key, weight) => {
  return {weight: weight, type: 'ed25519_public_key', key: key, public_key: key}
}

const mockServer = {
  loadAccount: jest.fn(account => {
    if (account === SIGNER.publicKey())
      return Promise.resolve(new sdk.Account(account, '1'))
    else if (account === JOINT_NEW.publicKey()) {
      const acc = new sdk.Account(account, '1')
      acc.signers = [
        signerObj(MEMBER1.publicKey(), 1),
        signerObj(MEMBER2.publicKey(), 1),
        signerObj(MEMBER3.publicKey(), 1),
      ]
      acc.thresholds = {
        low_threshold: 0,
        med_threshold: 0,
        high_threshold: MEMBERS.length,
      }
      return Promise.resolve(acc)
    }
  }),
  submitTransaction: jest.fn(() => Promise.resolve({fake: 'receipt'})),
}

// not using any network but the sdk needs the =to be set to something
sdk.Network.useTestNetwork()

const jointAccount = new JointAccount(sdk, mockServer)

it('joint account creates ok', async () => {
  const inputs = {
    accountSecret: JOINT_NEW.secret(),
    members: MEMBERS,
    signerSecret: SIGNER.secret(),
  }
  const receipt = await jointAccount.create(inputs)
  expect(receipt).toMatchSnapshot()
})
