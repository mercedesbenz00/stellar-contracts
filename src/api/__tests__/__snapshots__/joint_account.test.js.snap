// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joint account creates ok 1`] = `
Object {
  "inputs": Object {
    "accountSecret": "SABRXC3RQJVVUS22N7H6WGZPTWJ7GBKIZ3PBGX4NWKZZSLC4DWIUP4K3",
    "members": Array [
      "GAJ3VIHUHOHPI7HIJ3DS5HJ5K7CMJ5QVVCO4IXBIG32HGM3WAS3OEOMA",
      "GBN4BQAIFUEKYOEYAI4ZEWUK7UW4QVCFVT4VD643L5O7IBCZEES3R6D3",
      "GAGXSBXBSJUYCYM623OQQK5P3PXT2BOCKRXU7TVAEOZB5CET2ABJX2T6",
    ],
    "signerSecret": "SAP4WQ7W3JS72NGGTJ3X7FM3VGMVI5AIWS5VUOXK4MJECLTOPDUE7VMK",
  },
  "jointAccount": Object {
    "keys": Object {
      "publicKey": "GCKMUQMJXGSPMQRQAZZ2FTZBFCXIXWGWAKGCM3CUVTIR2OOZXB664RSO",
      "secret": "SABRXC3RQJVVUS22N7H6WGZPTWJ7GBKIZ3PBGX4NWKZZSLC4DWIUP4K3",
    },
    "signers": Array [
      Object {
        "key": "GAJ3VIHUHOHPI7HIJ3DS5HJ5K7CMJ5QVVCO4IXBIG32HGM3WAS3OEOMA",
        "public_key": "GAJ3VIHUHOHPI7HIJ3DS5HJ5K7CMJ5QVVCO4IXBIG32HGM3WAS3OEOMA",
        "type": "ed25519_public_key",
        "weight": 1,
      },
      Object {
        "key": "GBN4BQAIFUEKYOEYAI4ZEWUK7UW4QVCFVT4VD643L5O7IBCZEES3R6D3",
        "public_key": "GBN4BQAIFUEKYOEYAI4ZEWUK7UW4QVCFVT4VD643L5O7IBCZEES3R6D3",
        "type": "ed25519_public_key",
        "weight": 1,
      },
      Object {
        "key": "GAGXSBXBSJUYCYM623OQQK5P3PXT2BOCKRXU7TVAEOZB5CET2ABJX2T6",
        "public_key": "GAGXSBXBSJUYCYM623OQQK5P3PXT2BOCKRXU7TVAEOZB5CET2ABJX2T6",
        "type": "ed25519_public_key",
        "weight": 1,
      },
    ],
    "thresholds": Object {
      "high_threshold": 3,
      "low_threshold": 0,
      "med_threshold": 0,
    },
  },
  "transactions": Object {
    "createJointAccount": Object {
      "fake": "receipt",
    },
  },
}
`;
